### DISCUSSION TITLE
Integration with other VPNService apps

### DISCUSSION THREAD

**Discussion by emanuele-f** [emanuele-f/PCAPdroid#391] [Ideas]
In #379 came out the discussion about features which are currently spread across different firewall/network monitor apps. Since such apps often do not communicate (e.g. it's currently not possible to use both PCAPdroid and RethinkDNS at the same time), it's impossible for users to use them at the same time (because of the one VPNService limitation of Android).

This is meant to be an open discussion, allowing users expressing their needs and developers their point of view. To focus the objective, it will be useful to list which features PCAPdroid/other apps can provide to complement other apps/PCAPdroid functions.

- A solution that users can already use is work profiles and EveryProxy, as explained in https://github.com/emanuele-f/PCAPdroid/issues/274#issuecomment-1295293801 . However, this is not very convenient
- An initial proposal for integration was provided in https://github.com/emanuele-f/PCAPdroid/issues/274#issuecomment-1294126691 (read more on that task for more insights), to implement a new mode in PCAPdroid to expose a SOCKS5 server interface which other apps like RethinkDNS can use. That would be similar to the VPN-mode (e.g. it can block the traffic) but instead of reading/writing the packets from the VPNService fd, it would read/write them from the SOCKS5 server sockets. From a first analysis, this seems quite complex to implement. **NOTE**: this does not provide PCAPdroid the ability to use Wireguard/DNS over TLS, because PCAPdroid would be the last element in the chain. Instead, for this to work, RethinkDNS should provide the SOCKS5 server interface

Features that could complement PCAPdroid:
- DNS over HTTPS: https://github.com/emanuele-f/PCAPdroid/issues/211 (provided by InviZible, RethinkDNS, dnscrypt-proxy/android) . Update: it's now possible to integrate with InviZible Pro, see https://emanuele-f.github.io/PCAPdroid/advanced_features#42-doh-support
- Connect to Tor (provided by InviZible, Orbot, integrated in RethinkDNS), Update: See https://emanuele-f.github.io/PCAPdroid/advanced_features#47-tor for integration instructions
- Connect to Wireguard (provided by Wireguard, integrated in RethinkDNS) - see #397
- Connect to OpenVPN (provided by [ics-openvpn](https://github.com/schwabe/ics-openvpn#controlling-from-external-apps))

**Comment by anpic** [comment]
Thanks for understanding :)

> Since such apps often do not communicate

There is also a variant to participate in proxy chains and work in local proxy server mode. Although independent features doesn't interfere with this :)

> This is meant to be an open discussion, allowing users expressing their needs and developers their point of view. To focus the objective, it will be useful to list which features PCAPdroid/other apps can provide to complement other apps/PCAPdroid functions.

So we can discuss features here not only in the context of VpnService?

> DNS over TLS

It seems to me that the implementation of DnsCrypt (https://github.com/DNSCrypt/dnscrypt-proxy, https://github.com/quindecim/dnscrypt-proxy-android) can be no less useful ;)

> Connect to Wireguard (provided by Wireguard, integrated in RethinkDNS)

Wireguard is already implemented in Rethink, so it would be more interesting to think about integration with OpenVPN  (https://github.com/lawtancool/ics-openvpn-xor, https://github.com/IngoZenz/ics-openvpn, https://github.com/schwabe/ics-openvpn) ;)

> However, to provide a nice user experience, it's better to think like a dumb user than a tech savvy

We can discuss about that. But you've already come up with a working version yourself but you don't want to admit it in any way ;)
Let there be the simplest UI features for some users. But for smart users create a powerful system based on intents. When all functions can be controlled through intents. And even for those for whom you don't want to make UI elements ;)
It doesn't require any changes in UI. But it will attract the maximum number of users :)

**Comment by emanuele-f** [comment]
> There is also a variant to participate in proxy chains and work in local proxy server mode. Although independent features doesn't interfere with this :)

You mean to expose the SOCKS5 server interface, or is this something different? If so, do you have references for this other integration method?

>So we can discuss features here not only in the context of VpnService?

Yes. The objective is to discuss ways to remove barriers to let the users access all the features that they need, regardless of the app, either via inter-app communication or by direct integration

> It seems to me that the implementation of DnsCrypt (https://github.com/DNSCrypt/dnscrypt-proxy, https://github.com/quindecim/dnscrypt-proxy-android) can be no less useful ;)
> Wireguard is already implemented in Rethink, so it would be more interesting to think about integration with OpenVPN (https://github.com/lawtancool/ics-openvpn-xor, https://github.com/IngoZenz/ics-openvpn, https://github.com/schwabe/ics-openvpn) ;)

Thanks for pointing to them, added to the description above

> When all functions can be controlled through intents

The Intent API is meant to allow third-party apps to interface with PCAPdroid, mainly for traffic dump purposes. It's not meant to replace the UI, so making everything configurable via it does not make sense to me

**Reply by anpic** [reply]
> You mean to expose the SOCKS5 server interface

yeah

> mainly for traffic dump purposes

This doesn't mean that the purposes aren't extensible :)

> It's not meant to replace the UI

It's not only about replacing, it's about being a basic variant. Especially if the developer can't spend time developing ;)

> so making everything configurable via it does not make sense to me

Well, then how do you think to automate the work with android apps? Including PCAPdroid
Android Intents are a universal and the only correct means for this.
The userspace system is based on intents, and any really smart android project should be able to use their power and versatility.

**Reply by emanuele-f** [reply]
Without a real use-case, this would only take development time and increase code complexity

**Reply by anpic** [reply]
> this would only take development time and increase code complexity

You can say that about anything at all :)
Let's not develop anything else at all, because this would only take time and increase code complexity :))
It seemed to me that this topic appeared specifically so that it would be impossible to use this argument here ;)
It's clear that you're alone, and you don't owe anyone anything.
But this topic seems to be for user suggestions, and not for monosyllabic attempts to convince everyone that no one needs a full-fledged firewall.
You don't have any serious technical arguments against intents. Only a lack of time and desire ;)

**Comment by anpic** [comment]
> Yes. The objective is to discuss ways to remove barriers to let the users access all the features that they need, regardless of the app, either via inter-app communication or by direct integration

Okay, then I'll continue for future generations of developers ;)
I will gradually compare the missing features with the current android firewalls.
Let AFWall (https://github.com/ukanth/afwall) be the first.

- apps can be viewed, searched, sorted and log viewed by UIDs
<details>
  <summary>pics</summary>

![1](https://github.com/emanuele-f/PCAPdroid/assets/40465616/c2171b2f-ccea-40a0-8146-22a6ef6291f9) 
![7](https://github.com/emanuele-f/PCAPdroid/assets/40465616/e7507565-1ffc-4617-839b-953ce1be3385)
![2](https://github.com/emanuele-f/PCAPdroid/assets/40465616/400e2288-2f90-4014-b95c-ecf35bf01dd9)
![3](https://github.com/emanuele-f/PCAPdroid/assets/40465616/f99d5ce1-abd6-44ee-aa8a-6326bcd438a0)
</details>

- multiple profiles to quickly change the set of connection rules
<details>
  <summary>pics</summary>

![8](https://github.com/emanuele-f/PCAPdroid/assets/40465616/776e8cf7-8f8c-45f3-8f3d-eeb42dc4ca9d)
![9](https://github.com/emanuele-f/PCAPdroid/assets/40465616/722e1396-ba9d-425a-892f-64ba555f7e83)
</details>

- filtering connections by network interface
<details>
  <summary>pics</summary>

![4](https://github.com/emanuele-f/PCAPdroid/assets/40465616/ef59a992-002d-42b2-9c71-d88d1b256aba)
</details>

- the ability to hide app icons to speed up the loading of the app list
<details>
  <summary>pics</summary>

![5](https://github.com/emanuele-f/PCAPdroid/assets/40465616/6ff7c593-4786-4d8b-98ff-231273b36d4d)
</details>

- protection against arbitrary access and uninstalling
<details>
  <summary>pics</summary>

![6](https://github.com/emanuele-f/PCAPdroid/assets/40465616/936c395e-399d-491f-a1f7-88070900def7)
![11](https://github.com/emanuele-f/PCAPdroid/assets/40465616/45691d44-3509-462a-a135-3ba1d9e1b191)
</details>

- backup and restore all settings
<details>
  <summary>pics</summary>

![10](https://github.com/emanuele-f/PCAPdroid/assets/40465616/ab5cf69c-7c0e-44c6-b406-72eb915ca098)
</details>

**Comment by anpic** [comment]
Fiwalld non-root firewall (https://play.google.com/store/apps/details?id=com.lazybean.vfiwalld)
DoH Guard (https://play.google.com/store/apps/details?id=com.lazybean.doh)

- per app and ip:port settings
<details>
  <summary>pics</summary>

![1](https://github.com/emanuele-f/PCAPdroid/assets/40465616/8da415cd-275e-4d6b-932e-65aa62ca7780)
</details>

- per country settings and statistics

<details>
  <summary>pics</summary>

![2](https://github.com/emanuele-f/PCAPdroid/assets/40465616/1b8764df-25c4-4654-8a5a-47342de0d78a)
![3](https://github.com/emanuele-f/PCAPdroid/assets/40465616/c7a4f3cb-a63e-46c0-86ec-b7d97d2948ed)
</details>

- summary statistics
<details>
  <summary>pics</summary>

![4](https://github.com/emanuele-f/PCAPdroid/assets/40465616/0717564f-5f99-4e46-9544-1a8af80ff48b)
![5](https://github.com/emanuele-f/PCAPdroid/assets/40465616/f31548fc-2b18-4c5c-9410-231e1a3b4737)
![6](https://github.com/emanuele-f/PCAPdroid/assets/40465616/0d756f31-9656-44b6-81a9-ce800e2a63ba)
</details>

- choosing DNS servers
<details>
  <summary>pics</summary>

![7](https://github.com/emanuele-f/PCAPdroid/assets/40465616/54d529c7-afb6-4798-bd6e-47ce9a72b8bf)
![8](https://github.com/emanuele-f/PCAPdroid/assets/40465616/ba258d49-c2e5-4093-bcf9-fb03de41a067)
</details>

**Comment by anpic** [comment]
VProxid (https://play.google.com/store/apps/details?id=com.lazybean.vpnperapp)
RProxid (https://play.google.com/store/apps/details?id=com.lazybean.socksperapp)

- per app proxification with json import and export

<details>
  <summary>pics</summary>

![9](https://github.com/emanuele-f/PCAPdroid/assets/40465616/535b01c1-aba5-4efb-8554-35724e0926ac)
![10](https://github.com/emanuele-f/PCAPdroid/assets/40465616/cedc621b-26a8-4122-878d-6a1b9661a26d)
![11](https://github.com/emanuele-f/PCAPdroid/assets/40465616/633d9fb8-80ad-4052-9aa0-0815ea8ea54a)
</details>

**Comment by anpic** [comment]
NoRoot Firewall (https://play.google.com/store/apps/details?id=app.greyshirts.firewall)

- per app and [ip|domain]:port settings
<details>
  <summary>pics</summary>

![1](https://github.com/emanuele-f/PCAPdroid/assets/40465616/0e927d62-bd59-42e3-af5c-51cc0b5a1324)
![2](https://github.com/emanuele-f/PCAPdroid/assets/40465616/298586f5-515b-4ce1-b58d-4dde23850562)
</details>

- per app log
<details>
  <summary>pics</summary>

![3](https://github.com/emanuele-f/PCAPdroid/assets/40465616/97e3ec2b-7684-484a-89b4-b65f8ba5b327)
![4](https://github.com/emanuele-f/PCAPdroid/assets/40465616/1c011bef-8294-4a1e-b6ad-7b102c467161)
</details>

- globbing
<details>
  <summary>pics</summary>

![5](https://github.com/emanuele-f/PCAPdroid/assets/40465616/b1324cc7-7a93-4109-85c4-c64992fa696e)
![6](https://github.com/emanuele-f/PCAPdroid/assets/40465616/f0368219-2192-47eb-9858-091bb25dbde7)
</details>

- full backup and restore in json
<details>
  <summary>pics</summary>

![7](https://github.com/emanuele-f/PCAPdroid/assets/40465616/524b4d60-6c27-4658-b923-8bb68b3141f4)
</details>

**Comment by anpic** [comment]
NetPatch Firewall (https://github.com/netpatch/firewall)

- per app rules

<details>
  <summary>pics</summary>

![1](https://github.com/emanuele-f/PCAPdroid/assets/40465616/84c61240-7767-473c-957b-4588dbc4dfc2)

![2](https://github.com/emanuele-f/PCAPdroid/assets/40465616/595a48b4-ceee-45bc-9aa9-27dae42acf06)
</details>

- domain and ip groups

<details>
  <summary>pics</summary>

![3](https://github.com/emanuele-f/PCAPdroid/assets/40465616/9e3cfb39-ab01-483f-ad3a-d31c30c6b938)

![4](https://github.com/emanuele-f/PCAPdroid/assets/40465616/15d2e384-ed3f-4b7b-9e50-c4d8b745be4b)
</details>

- per app statistics

<details>
  <summary>pics</summary>

![5](https://github.com/emanuele-f/PCAPdroid/assets/40465616/d10d633f-8219-4a80-9a3b-21f00914dae1)
</details>

- Shadowsocks proxy

<details>
  <summary>pics</summary>

![6](https://github.com/emanuele-f/PCAPdroid/assets/40465616/9be69538-4dcc-4aae-9bc5-8a09bcb7357f)
</details>

- DNS through proxy, leak prevention, cache time

<details>
  <summary>pics</summary>

![7](https://github.com/emanuele-f/PCAPdroid/assets/40465616/febf364b-cf8e-4889-90a8-af9c140e2764)

![8](https://github.com/emanuele-f/PCAPdroid/assets/40465616/7ad629f9-6e2e-4067-a8b9-36004c058d52)
</details>

- globbing

<details>
  <summary>pics</summary>

![9](https://github.com/emanuele-f/PCAPdroid/assets/40465616/529cfa8e-c721-465f-b1c3-1d4de310afa2)
</details>

- backup and restore all settings
